src/openads-appnexus/repository/PullingAdRepository.js

```diff
@@ -10,44 +12,58 @@ export default class PullingAdRepository {
   }
 
   find ({id}) {
-    return Promise.race([
-      this._waitForData({id}),
-      this._timeoutPromise()
-    ])
+    return Promise.resolve(id)
+      .then(id => this._ads.has(id) ? this._ads.get(id) : this._createPullingDataEntry(id))
+      .then(pullingDataEntry => Promise.race([
+        this._waitForData(pullingDataEntry),
+        this._timeoutPromise(pullingDataEntry)
+      ]))
   }
 
   has ({id}) {
-    return this._ads.has(id)
+    return this._ads.has(id) && this._ads.get(id).data !== undefined
   }
 
   save ({id, adResponse}) {
-    this._ads.set(id, adResponse)
+    if (!this._ads.has(id)) {
+      this._ads.set(id, new PullingDataEntry(id))
+    }
+    this._ads.get(id).updateData(adResponse)
   }
 
   remove ({id}) {
     return this._ads.delete(id)
   }
 
-  _waitForData ({id}) {
-    return Promise.resolve(this._ads.get(id))
-      .then(optionalAd => optionalAd || this._intervalPull(id))
+  _createPullingDataEntry (id) {
+    const pullingDataEntry = new PullingDataEntry(id)
```

Comment:
As we have spoken already, the concept of having a PullingDataEntry to store information related with ad data and id's intervals is a great idea.
We should work around to this idea and encapsulate responsabilities ;)